File: STACMR-matlab/staCMRFITBN.m
function [p, datafit, fits, pars] = staCMRFITBN (data, varargin)
% [p, datafit, fits, pars] = staCMRFITBN (data, varargin)
% Binomial version of staCMRFIT
% nsample = no. of Monte Carlo samples (about 10000 is good)
% data = data structure (cell array, general, or structured)
% model is a nvar * k matrix specifying the linear model; default = ones(nvar,1)
% E = optional partial order model e.g. E={[1 2] [3 4 5]} indicates that
% condition 1 <= condition 2 and condition 3 <= condition 4 <= condition 5
%
% returns:
% p = empirical p-value
% datafit = observed fit of monotonic (1D) model
% fits = nsample vector of fits of Monte Carlo samples (it is against this
% distribution that datafit is compared to calculate p)
% % Note: These are g-squared values (not least squares)
% % pars = nvar cell array of bootstrap means nsample x ncond x nsub
% *************************************************************************
% Last modified: 24 April 2017
% pars output added 4 March 2020
% *************************************************************************
%
if ~iscell(data)
    y = BNgen2cell(data);
else
    y = data;
end

% defaults
nsample = 1;
E = {};
nvar = size(y,2);
model = ones(nvar,1);
approximate = 0;
showStatus = 0;
ranseed = -1;
tol = 1e-6;

% set arguments if specified
for i = 1 : 2 : length(varargin)-1
    name = varargin{i}; 
    value = varargin{i+1};
    switch name
        case {'nsamples', 'nsample', 'ns', 'n'} 
            nsample = value; % number of samples
        case {'partial', 'part', 'p'}
            E = value; % partial order
        case {'model','mod','m'}
            model = value; % model
        case {'approximate','approx','a'}
            approximate = value; % approximate solution if true
        case {'showstatus','show','s'}
            showStatus = value; % Luke's update window
        case {'ranseed','ran','r'}
            ranseed = value; % random number seed
    end
end

if ~iscell(E)
    E = adj2cell(E); % convert from adjacency matrix form
end

[p, datafit, fits, pararray] = jCMRxBNfits(nsample, y, E, model, approximate, showStatus, ranseed);
% additional changes March 2020
datafit(datafit<tol) = 0;
fits (fits < tol) = 0;
pars = cell(1,nvar); % pars contains bootstrap means
for ivar=1:nvar
	z = squeeze(pararray(:,:,ivar,:));
    pars{ivar} = permute(z, [1 3 2]);
end

File: STACMR-R/staCMRFITBN.R
staCMRFITBN <- function (data=NULL, partial = list(), nsample=1, model=NULL, proc=-1, approx=0) {
  # input:
  # nsample = no. of Monte Carlo samples (about 10000 is good)
  # data = data structure (cell array or general)
  # model is a nvar * k matrix specifying the linear model, default = ones(nvar,1))
  # partial = optional partial order model e.g. E={[1 2] [3 4 5]} indicates that
  # condition 1 <= condition 2 and condition 3 <= condition 4 <= condition 5
  # default = none (empty)
  
  # output:
  # p = empirical p-value
  # datafit = observed fit of monotonic (1D) model
  # fits = nsample vector of fits of Monte Carlo samples (it is against this
  # distribution that datafit is compared to calculate p)
  # *************************************************************************
  # converted from matlab 7 February 2018
  # *************************************************************************
  
  if (is(data,"data.frame")) {
    y = gen2listBN (data) # convert from general format to list format
  } else {y = data} 
  
  a = "means" %in% variable.names(as.data.frame(y))
  if (!a) {y = binSTATS(y)}
  
  tol <- 10e-6
  if (missing(partial)) {partial = list()}
  
  nvar =length(y)
  model = NULL
  # if (missing(model) | is.null(model)) {model = matrix(1,nvar,1)} # sta default model
  
  if (!is.list(partial)) {partial = adj2list(partial)} # convert from adjacency matrix to list
  
  output = jCMRxBNfits(nsample, y, partial, model, proc, approx) # call java program
  
  output$fits[which(output$fits <= tol)] = 0;
  
  return (output)
}